src/core/regimes.rkt

```diff
@@ -25,16 +25,42 @@
 ;; `infer-splitpoints` and `combine-alts` are split so the mainloop
 ;; can insert a timeline break between them.
 
-(define (infer-splitpoints alts ctx)
-  (timeline-event! 'regimes)
+(define (infer-splitpoints alts branch-exprs cerrs cbest-index ctx)
   (timeline-push! 'inputs (map (compose ~a alt-expr) alts))
-  (define branch-exprs
-    (if (flag-set? 'reduce 'branch-expressions)
-        (exprs-to-branch-on alts ctx)
-        (context-vars ctx)))
   (define err-lsts (batch-errors (map alt-expr alts) (*pcontext*) ctx))
-  (define options (for/list ([bexpr branch-exprs]) (option-on-expr alts err-lsts bexpr ctx)))
-  (define best (argmin (compose errors-score option-errors) options))
+
+  ;; we want to try the one we picked last time
+  ;; and then if there are others in the cerrs that are better, try it.
+  (define init 
+    (cond 
+      [(= +inf.0 (list-ref cerrs cbest-index)) null]
+      [else (option-on-expr alts err-lsts (list-ref branch-exprs cbest-index) ctx)]))
+  (define init-err 
+    (cond 
+      [(null? init) +inf.0] 
+      [else (errors-score (option-errors init))]))
+  (define init-index cbest-index)
+
+  ;; invariant:
+  ;; errs[i] is some best option on branch-expr[i] computed on more alts than we have right now.
+  ;;
+  ;; errs[i] will be +inf.0 if we recomputed the branch expressions for the current set of alts
+  ;; and branch-expr[i] no longer exists.
+  (match-define-values (best best-err best-index errs) 
+      (for/fold ([best init] [best-err init-err] [best-index init-index] [errs (list-set cerrs init-index init-err)] 
+            #:result (values best best-err best-index errs)) 
+            ([bexpr branch-exprs] [berr cerrs] [i (range (length branch-exprs))])
+    ;; don't recompute if we know we've computed this branch-expr on more alts and it's still worse
+    (cond 
+      [(and (< berr best-err) (not (= i cbest-index)))
+        (define opt (option-on-expr alts err-lsts bexpr ctx))
+        (define err (errors-score (option-errors opt)))
```

Comment:
This could be a different PR but I kind of think we should just bite the bullet and add the number of branches to the `errors-score` here, so that the invariants assumed by branch-and-bound were satisfied.